src/main/java/mops/businesslogic/DirectoryServiceImpl.java

```diff
@@ -0,0 +1,253 @@
+
+    /**
+     * Represents the role of an admin.
+     */
+    public static final String ADMINISTRATOR = "administrator";
+    /**
+     * This connects to database related to directory information.
+     */
+    private final DirectoryRepository directoryRepository;
+
+    /**
+     * API for GruppenFindung which handles permissions.
+     */
+    private final PermissionService permissionService;
+
+    /**
+     * Handles meta data of files.
+     */
+    private final FileInfoService fileInfoService;
+
+    /**
+     * Handle permission checks for roles.
+     */
+    private final RoleServiceImpl roleService;
+
+    /**
+     * This connects to database to handle directory permissions.
+     */
+    private final DirectoryPermissionsRepository directoryPermissionsRepo;
+
+    /**
+     * Checks whether the user is authorized to load the file.
+     *
+     * @param account user credentials
+     * @param dirId   the id of the folder where the file will be uploaded
+     */
+    @Override
+    public void checkWritePermission(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+
+        roleService.checkWritePermission(account, directory);
+    }
+
+    /**
+     * Checks if user has reading permission.
+     *
+     * @param account user credentials
+     * @param dirId   the id of the folder
+     */
+    @Override
+    public void checkReadPermission(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+        roleService.checkReadPermission(account, directory);
+    }
+
+    /**
+     * Checks if user has deleting permission.
+     *
+     * @param account user credentials
+     * @param dirId   the id of the folder
+     */
+    @Override
+    public void checkDeletePermission(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+        roleService.checkDeletePermission(account, directory);
+    }
+
+    /**
+     * Returns all folders of the parent folder.
+     *
+     * @param account     user credentials
+     * @param parentDirID id of the parent folder
+     * @return list of folders
+     */
+    @Override
+    public List<Directory> getSubFolders(Account account, long parentDirID) throws MopsException {
+        Directory directory = fetchDirectory(parentDirID);
+        roleService.checkReadPermission(account, directory);
+        return directoryRepository.getAllSubFoldersOfParent(parentDirID);
+    }
+
+
+    /**
+     * Creates the group root directory.
+     *
+     * @param account user credentials
+     * @param groupId the group id
+     * @return the directory created
+     */
+    @Override
+    public Directory createRootFolder(Account account, Long groupId) throws MopsException {
+
+        roleService.checkIfRole(account, groupId, ADMINISTRATOR);
+        Set<String> roleNames = permissionService.fetchRolesInGroup(groupId);
```

Comment:
This could be moved to `RoleService`. It would have the advantage that `DirectoryService` will no longer call `PermissionService`. But for one line it could be overkill. What do you think?

Reply:
I think for a one-time occurance it's fine to leave it this way.

Reply:
It would be splendid if we could move everything not about directories (roles, permissions etc.) somewhere else

Reply:
> It would be splendid if we could move everything not about directories (roles, permissions etc.) somewhere else

Then I will do it.